packages/discord.js/typings/index.d.ts

```diff
@@ -3207,28 +3207,28 @@ export class ThreadChannel<ThreadOnly extends boolean = boolean> extends BaseCha
   public type: ThreadChannelType;
   public get unarchivable(): boolean;
   public delete(reason?: string): Promise<this>;
-  public edit(options: ThreadEditOptions): Promise<AnyThreadChannel>;
-  public join(): Promise<AnyThreadChannel>;
-  public leave(): Promise<AnyThreadChannel>;
+  public edit(options: ThreadEditOptions): Promise<this>;
+  public join(): Promise<this>;
+  public leave(): Promise<this>;
   public permissionsFor(memberOrRole: GuildMember | Role, checkAdmin?: boolean): Readonly<PermissionsBitField>;
   public permissionsFor(
     memberOrRole: GuildMemberResolvable | RoleResolvable,
     checkAdmin?: boolean,
   ): Readonly<PermissionsBitField> | null;
   public fetchOwner(options?: BaseFetchOptions): Promise<ThreadMember | null>;
   public fetchStarterMessage(options?: BaseFetchOptions): Promise<Message<true> | null>;
-  public setArchived(archived?: boolean, reason?: string): Promise<AnyThreadChannel>;
+  public setArchived(archived?: boolean, reason?: string): Promise<this>;
   public setAutoArchiveDuration(
     autoArchiveDuration: ThreadAutoArchiveDuration,
     reason?: string,
-  ): Promise<AnyThreadChannel>;
-  public setInvitable(invitable?: boolean, reason?: string): Promise<AnyThreadChannel>;
-  public setLocked(locked?: boolean, reason?: string): Promise<AnyThreadChannel>;
-  public setName(name: string, reason?: string): Promise<AnyThreadChannel>;
+  ): Promise<this>;
+  public setInvitable(invitable?: boolean, reason?: string): Promise<this>;
+  public setLocked(locked?: boolean, reason?: string): Promise<this>;
+  public setName(name: string, reason?: string): Promise<this>;
   // The following 3 methods can only be run on forum threads.
-  public setAppliedTags(appliedTags: readonly Snowflake[], reason?: string): Promise<ThreadChannel<true>>;
-  public pin(reason?: string): Promise<ThreadChannel<true>>;
-  public unpin(reason?: string): Promise<ThreadChannel<true>>;
+  public setAppliedTags(appliedTags: readonly Snowflake[], reason?: string): Promise<If<ThreadOnly, this, never>>;
+  public pin(reason?: string): Promise<If<ThreadOnly, this, never>>;
+  public unpin(reason?: string): Promise<If<ThreadOnly, this, never>>;
```

Comment:
Doesn't this assert the type based on... if an error is thrown? 👀

Reply:
Yeah. These methods exist at runtime on this class, so it'd be incorrect to move them elsewhere, and we couldn't move them in the code since there aren't any other classes, just interfaces. However, they only work in forum threads, as noted in the comment. They are guaranteed to throw elsewhere. There's no good way to restrict usage of these outside forum threads, but I figured we might as well flag the possible mistake in the return value.
Typically TypeScript recognizes functions that return `undefined` and flags unreachable code after them. Unfortunately, it doesn't yet do that for asynchronous functions, at least currently. But it still doesn't hurt to add it.
![image](https://github.com/discordjs/discord.js/assets/61329810/c1744578-4b3f-44b4-9dc2-8429f39cc116)
![image](https://github.com/discordjs/discord.js/assets/61329810/fd889e14-b336-4d3e-a5ff-b0261f36a6ec)

Also, the return type will be `this` even if `ThreadOnly` is just `boolean`. `never` is removed from unions, so the return type will only be `never` if it's guaranteed to be a non-forum thread throw.

Reply:
These functions aren't actually marked as async themselves, they just return the promise from `ThreadChannel#edit`. So if we wanted to, we could add our own runtime check for these methods (currently it just relies on the Discord API to fail). Then we could also change this return type to `If<ThreadOnly, Promise<this>, never>`, allowing TS's unreachability detection to work here.

Reply:
could we overwtite the functions at runtime. Is there something in the jsdoc as well to advise of this.